src/math_functions.cpp

```diff
@@ -0,0 +1,747 @@
+  // following recurrence relations so that only a single sin/cos have to be
+  // evaluated (http://mathworld.wolfram.com/Multiple-AngleFormulas.html)
+  //
+  // sin(nx) = 2 cos(x) sin((n-1)x) - sin((n-2)x)
+  // cos(nx) = 2 cos(x) cos((n-1)x) - cos((n-2)x)
+
+  double sin_phi;
+  double cos_phi;
+  sin_phi = std::sin(phi);
+  cos_phi = std::cos(phi);
+
+  double sin_phi_vec[n + 1]; // Sin[n * phi]
+  double cos_phi_vec[n + 1]; // Cos[n * phi]
+  sin_phi_vec[0] = 1.0;
+  cos_phi_vec[0] = 1.0;
+  sin_phi_vec[1] = 2.0 * cos_phi;
+  cos_phi_vec[1] = cos_phi;
+
+  for (int i = 2; i <= n; i++) {
+    sin_phi_vec[i] = 2. * cos_phi * sin_phi_vec[i - 1] - sin_phi_vec[i - 2];
+    cos_phi_vec[i] = 2. * cos_phi * cos_phi_vec[i - 1] - cos_phi_vec[i - 2];
+  }
+
+  for (int i = 0; i <= n; i++) {
+    sin_phi_vec[i] *=  sin_phi;
+  }
+
+  // ===========================================================================
+  // Calculate R_pq(rho)
+  double zn_mat[n + 1][n + 1]; // Matrix forms of the coefficients which are
+                               // easier to work with
+
+  // Fill the main diagonal first (Eq 3.9 in Chong)
+  for (int p = 0; p <= n; p++) {
+    zn_mat[p][p] = std::pow(rho, p);
+  }
+
+  // Fill the 2nd diagonal (Eq 3.10 in Chong)
+  for (int q = 0; q <= n - 2; q++) {
+    zn_mat[q][q+2] = (q + 2) * zn_mat[q+2][q+2] - (q + 1) * zn_mat[q][q];
+  }
+
+  // Fill in the rest of the values using the original results (Eq. 3.8 in Chong)
+  double k1;
+  double k2;
+  double k3;
+  double k4;
+  for (int p = 4; p <= n; p++) {
+    k2 = static_cast<double> (2 * p * (p - 1) * (p - 2));
+    for (int q = p - 4; q >= 0; q -= 2) {
+      k1 = static_cast<double>((p + q) * (p - q) * (p - 2)) / 2.;
+      k3 = static_cast<double>(-q * q * (p - 1) - p * (p - 1) * (p - 2));
+      k4 = static_cast<double>(-p * (p + q - 2) * (p - q - 2)) / 2.;
+      zn_mat[q][p] =
+        ((k2 * rho * rho + k3) * zn_mat[q][p-2] + k4 * zn_mat[q][p-4]) / k1;
+    }
+  }
+
+  // Roll into a single vector for easier computation later
+  // The vector is ordered (0,0), (1,-1), (1,1), (2,-2), (2,0),
+  // (2, 2), ....   in (n,m) indices
+  // Note that the cos and sin vectors are offset by one
+  // sin_phi_vec = [sin(x), sin(2x), sin(3x) ...]
+  // cos_phi_vec = [1.0, cos(x), cos(2x)... ]
+  int i;
+  i = 0;
+  for (int p = 0; p <= n; p++) {
+    for (int q = -p; q <= p; q += 2) {
+      if (q < 0) {
+        zn[i] = zn_mat[std::abs(q)][p] * sin_phi_vec[std::abs(q) - 1];
+      } else if (q == 0) {
+        zn[i] = zn_mat[q][p];
+      } else {
+        zn[i] = zn_mat[q][p] * cos_phi_vec[q];
+      }
+      i++;
+    }
+  }
+
+}
+
+
+void rotate_angle_c(double uvw[3], const double mu, double* phi) {
+  // Copy original directional cosines
+  double u0;     // original cosine in x direction
+  double v0;     // original cosine in y direction
+  double w0;     // original cosine in z direction
+
+  u0 = uvw[0];
+  v0 = uvw[1];
+  w0 = uvw[2];
```

Comment:
Ok.  Will do; I went with this way as I considered it a good practice to do since doing a multi-line declare and assign with a class would call the default constructor followed by the `=` operator where as doing it on one-line calls the copy constructor.  I figured consistency would be good, even if those constructors dont matter for doubles.  However, I'm not really stuck to that.

- [x]